cpp/src/arrow/io/readahead.h

```diff
@@ -0,0 +1,94 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_IO_READAHEAD_H
+#define ARROW_IO_READAHEAD_H
+
+#include <memory>
+#include <string>
+
+#include "arrow/io/interfaces.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class MemoryPool;
+class ResizableBuffer;
+class Status;
+
+namespace io {
+namespace internal {
+
+struct ARROW_EXPORT ReadaheadBuffer {
+  std::shared_ptr<ResizableBuffer> buffer;
+  int64_t left_padding;
+  int64_t right_padding;
+};
+
+class ARROW_EXPORT ReadaheadSpooler {
+ public:
+  /// \brief EXPERIMENTAL: Create a readahead spooler wrapping the given input stream.
+  ///
+  /// The spooler launches a background thread that reads up to a given number
+  /// of fixed-size blocks in advance from the underlying stream.
```

Comment:
Could use a bit more documentation about what left padding and right padding are

Reply:
Done.